lightning/src/ln/channel.rs

```diff
@@ -10422,11 +10557,115 @@ where
 
 		// Note on channel reserve requirement pre-check: as the splice acceptor does not contribute,
 		// it can't go below reserve, therefore no pre-check is done here.
-		// TODO(splicing): Once splice acceptor can contribute, add reserve pre-check, similar to the one in `splice_ack`.
 
-		// TODO(splicing): Store msg.funding_pubkey
-		// TODO(splicing): Apply start of splice (splice_start)
+		let pre_channel_value = self.funding.value_to_self_msat;
+		let _post_channel_value = PendingSplice::compute_post_value(
+			pre_channel_value,
+			their_funding_contribution_satoshis,
+			our_funding_contribution_satoshis,
+		);
+		let _post_balance = PendingSplice::add_checked(
+			self.funding.value_to_self_msat,
+			our_funding_contribution_satoshis,
+		);
+		// TODO: Early check for reserve requirement
+
+		Ok(())
+	}
+
+	/// See also [`validate_splice_init`]
+	#[cfg(splicing)]
+	pub(crate) fn splice_init<ES: Deref, L: Deref>(
+		&mut self, msg: &msgs::SpliceInit, our_funding_contribution: i64, signer_provider: &SP,
+		entropy_source: &ES, holder_node_id: &PublicKey, logger: &L,
+	) -> Result<msgs::SpliceAck, ChannelError>
+	where
+		ES::Target: EntropySource,
+		L::Target: Logger,
+	{
+		let _res = self.validate_splice_init(msg)?;
+
+		let pre_channel_value = self.funding.get_value_satoshis();
+		let their_funding_contribution = msg.funding_contribution_satoshis;
 
+		let post_channel_value = PendingSplice::compute_post_value(
+			pre_channel_value,
+			our_funding_contribution,
+			their_funding_contribution,
+		);
+
+		let (our_funding_satoshis, their_funding_satoshis) = calculate_total_funding_contribution(
+			pre_channel_value,
+			our_funding_contribution,
+			msg.funding_contribution_satoshis,
+			false, // is_outbound
+		)?;
```

Comment:
Could any of this be moved into `FundingScope::for_splice`? It would be nice if we could avoid the `debug_assert` in there.

Reply:
Not really. The debug assert in `FundingScope::for_splice` says that we cannot splice out more than our balance. The reserve check should implicitly guard against this as well: we cannot splice out more than our balance plus the minimum reserve. (#3641).
The `calculate_total_funding_contribution` does not deal with current balances, so that's a different check.